RULES.md

```diff
@@ -0,0 +1,43 @@
+
+
+Ben's rules for writing performant database applications
+
+These are lessons that I have learned over a decade of writing Django and other
+Python web applications dealing with large amounts of data.  They might be
+unusual and controversial, so that's why I am wrote them down here.  They are
+not rules that we must follow in this project.  They are a guide for if and
+when we need to scale to large amounts of data or have fast database
+operations.
+
+1. Think in sets and do bulk operations.  A single entry is a set of one.   The
+   cost of a single insert or update is similar to the cost of a bulk insert or
+   update since most of the cost of database writes are updating indicies.
+
+2. Know what SQL statement that you want to send to the database first.  Use
+   tools to generate that SQL second. ORMs generate really slow SQL unless you
```

Comment:
> ORMs generate really slow SQL

I disagree with this conclusion. People produce bad solutions unless they know exactly what they're doing. It's a matter of a competence, not a technology. 

However  Django and DRF experience may be harmful and makes the situation worse, due to Django's poor architecture. Django ORM gives a false impression of working with object model entirely and moves the focus away from the resulting SQL the ORM produces. DRF violates SRP and provides leaky abstractions (i.e. Serializers and Viewsets).

However we use the SQLAlchemy in the project, which is way closer to native SQL (unless used with implicit relationship loading techniques, that are not supported for asynchronous mode anyway). With SQLAlchemy 2.0 API the edge between Core and ORM becomes even thinner.
SQLAlchemy is simply a query builder, data mapper and a unit of work patterns, and I don't think there is anything wrong  in using them. In any more or less complicated project you'll end up building a data mapper on your own, unless taking an existing one.

I would put it differently:

Think about what you're doing. Think about number and complexity of queries you're writing. When deciding between processing data in Python application and the database, follow the advice nr. 5.